src/query/search.ts

```diff
@@ -0,0 +1,143 @@
+ * this will result in the same underlying searches being performed each time.
+ *
+ * Warning:
+ * Weaviate Agents - Query Agent is an early stage alpha product.
+ * The API is subject to breaking changes. Please ensure you are using the latest version of the client.
+ *
+ * For more information, see the [Weaviate Query Agent Docs](https://weaviate.io/developers/agents/query)
+ */
+export class QueryAgentSearcher<T> {
+  private agentsHost: string;
+  private query: string;
+  private collections: (string | QueryAgentCollectionConfig)[];
+  private systemPrompt?: string;
+  private cachedSearches?: ApiSearchResult[];
+
+  constructor(
+    private client: WeaviateClient,
+    query: string,
+    {
+      collections = [],
+      systemPrompt,
+      agentsHost = "https://api.agents.weaviate.io",
+    }: {
+      collections?: (string | QueryAgentCollectionConfig)[];
+      systemPrompt?: string;
+      agentsHost?: string;
+    } = {},
+  ) {
+    this.query = query;
+    this.collections = collections;
+    this.systemPrompt = systemPrompt;
+    this.agentsHost = agentsHost;
+    this.cachedSearches = undefined;
+  }
+
+  private async getHeaders() {
+    const { host, bearerToken, headers } =
+      await this.client.getConnectionDetails();
+    const requestHeaders = {
+      "Content-Type": "application/json",
+      Authorization: bearerToken!,
+      "X-Weaviate-Cluster-Url": host,
+      "X-Agent-Request-Origin": "typescript-client",
+    };
+    const connectionHeaders = headers;
+    return { requestHeaders, connectionHeaders };
+  }
+
+  private buildRequestBody(
+    limit: number,
+    offset: number,
+    connectionHeaders: HeadersInit | undefined,
+  ) {
+    const base = {
+      headers: connectionHeaders,
+      original_query: this.query,
+      collections: mapCollections(this.collections),
+      limit,
+      offset,
+    } as const;
+    if (this.cachedSearches === undefined) {
+      return {
+        ...base,
+        searches: null,
+        system_prompt: this.systemPrompt || null,
+      };
+    }
+    return {
+      ...base,
+      searches: this.cachedSearches,
+    };
+  }
+
+  /**
+   * Run the search-only agent with the given limit and offset values.
+   *
+   * Calling this method multiple times on the same QueryAgentSearcher instance will result
+   * in the same underlying searches being performed each time, allowing you to paginate
+   * over a consistent results set.
+   *
+   * @param options - Options for executing the search
+   * @param options.limit - The maximum number of results to return. Defaults to 20 if not specified.
+   * @param options.offset - The offset to start from. If not specified, retrieval begins from the first object.
+   * @returns A SearchModeResponse object containing the results, usage, and underlying searches performed.
+   */
+  async run(options: SearchExecutionOptions): Promise<SearchModeResponse<T>> {
```

Comment:
Should a user be able to just call `await search.run()`? If so, `options` needs to be optional here: `options?: SearchExecutionOptions`